### PR TITLE
integrated Jest for front-end unit testing

### PR DESCRIPTION
**Pull request by srizvi-official**
In this PR, we have integrated Jest as a front-end testing tool. Jest is developed and used by Facebook (the creators of React as well) and has great documentation support on their [official site](https://facebook.github.io/jest/). We have setup Jest to test any Javascript code including [React components](https://facebook.github.io/jest/docs/tutorial-react.html#content) and to use jQuery for [DOM manipulation](https://facebook.github.io/jest/docs/tutorial-jquery.htmll). Enzyme dependency for Jest is used to assert, and manipulate your rendered components. 
- You can write test cases in <RootDir>/app/javascript/__tests__ . See the examples there.
- If you are testing a module that requires any dependency on runtime like jquery, then require jest-setup.js from <RootDir>/config/
- If you write any test case, just run 'yarn jest' command to execute all test cases. If you want to run specific test file then execute this command 'yarn jest test-file-name'
- Don't forget to run npm install before starting anything

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,11 +24,19 @@
...
     "sass-loader": "^6.0.3",
+    "scripts": {
+      "test": "jest"
+    },
     "style-loader": "^0.16.1",
     "webpack": "^2.2.1",
     "webpack-merge": "^4.1.0"
   },
   "devDependencies": {
+    "babel-jest": "^19.0.0",
```
Comment (mckinley) [inline package.json]: We cant add to devDependencies like this. Please use `yarn` for asset dependency management. `yarn add ...`

File: app/javascript/sum.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+function sum(a, b) {
```
Comment (mckinley) [inline app/javascript/sum.js]: I appreciate that you added an example. but I don't think this should be a part of our codebase.

File: app/javascript/react/CheckboxWithLabel.jsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+
+export default class CheckboxWithLabel extends React.Component {
```
Comment (mckinley) [inline app/javascript/react/CheckboxWithLabel.jsx]: This is nice but I don't think this sort of thing should be in our codebase until it is needed.

Comment (srizvi-official) [inline app/javascript/react/CheckboxWithLabel.jsx]: Yes, you are right. But as we don't have much JS code to test, I added some to test the testing tool and its working. I am going to remove the sample tests in my next commit.

File: package.json (hunk 1 above)
Comment (srizvi-official) [inline package.json]: Now, added dependencies through yarn.

**Comment by mckinley** [comment]
I feel like there is a lot that still needs to happen here.

We are using webpack and according to documentation (https://facebook.github.io/jest/docs/webpack.html) there is a bit of configuration that needs to be done for other assets. We can get to those as we need them though I guess.

It should also be run as a part of rake.

Also we should not have any devDependencies other than the webpack dev server. I was not even able to `yarn install` because jest is listed twice now in the package.json file.

I feel like it might be easier to just redo this, starting from master. All it is so far is a few `yarn add`s and the included test (which could be copied over). If I have time tonight I will do that, otherwise you can pick it back up. At least please sort out the yarn packages.

**Comment by srizvi-official** [comment]
We will do that in separate branch as suggested.

**Comment by mckinley** [comment]
Thanks!

**Pull request closed without merging**